hathor/indexes/memory_deps_index.py

```diff
@@ -0,0 +1,170 @@
+            return
+        # add direct deps
+        if __debug__ and tx.hash in self._dir_dep_index:
+            # XXX: dependencies set must be immutable
+            assert self._dir_dep_index[tx.hash].keys() == _deps.keys()
+        self._dir_dep_index[tx.hash] = _deps
+        # add reverse dep
+        for rev_dep in _deps:
+            if rev_dep not in self._rev_dep_index:
+                self._rev_dep_index[rev_dep] = set()
+            self._rev_dep_index[rev_dep].add(tx.hash)
+
+    def _del_from_deps_index(self, tx: bytes) -> None:
+        _deps = self._dir_dep_index.pop(tx, _DirDepValue())
+        for rev_dep in _deps.keys():
+            rev_deps = self._rev_dep_index[rev_dep]
+            if tx in rev_deps:
+                rev_deps.remove(tx)
+            if not rev_deps:
+                del self._rev_dep_index[rev_dep]
+
+    def remove_ready_for_validation(self, tx: bytes) -> None:
+        """ Removes from ready for validation set.
+        """
+        self._txs_with_deps_ready.discard(tx)
+
+    def next_ready_for_validation(self, tx_storage: 'TransactionStorage', *, dry_run: bool = False) -> Iterator[bytes]:
+        if dry_run:
+            cur_ready = self._txs_with_deps_ready.copy()
+        else:
+            cur_ready, self._txs_with_deps_ready = self._txs_with_deps_ready, set()
+        while cur_ready:
+            yield from sorted(cur_ready, key=lambda tx_hash: tx_storage.get_transaction(tx_hash).timestamp)
+            if dry_run:
+                cur_ready = self._txs_with_deps_ready - cur_ready
+            else:
+                cur_ready, self._txs_with_deps_ready = self._txs_with_deps_ready, set()
+
+    def iter(self) -> Iterator[bytes]:
+        yield from self._rev_dep_index.keys()
+
+    def _get_rev_deps(self, tx: bytes) -> FrozenSet[bytes]:
+        """Get all txs that depend on the given tx (i.e. its reverse depdendencies)."""
+        return frozenset(self._rev_dep_index.get(tx, set()))
+
+    def known_children(self, tx: BaseTransaction) -> List[bytes]:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        it_rev_deps = map(tx.storage.get_transaction, self._get_rev_deps(tx.hash))
+        return [not_none(rev.hash) for rev in it_rev_deps if tx.hash in rev.parents]
+
+    # needed-txs-index methods:
+
+    def has_needed_tx(self) -> bool:
+        return bool(self._needed_txs_index)
+
+    def is_tx_needed(self, tx: bytes) -> bool:
+        return tx in self._needed_txs_index
+
+    def remove_from_needed_index(self, tx: bytes) -> None:
+        self._needed_txs_index.pop(tx, None)
+
+    def get_next_needed_tx(self) -> bytes:
+        # This strategy maximizes the chance to download multiple txs on the same stream
+        # find the tx with highest "height"
+        # XXX: we could cache this onto `needed_txs` so we don't have to fetch txs every time
+        height, start_hash, tx = max((h, s, t) for t, (h, s) in self._needed_txs_index.items())
```

Comment:
O(n)? Maybe we should use a separate data structure to keep priorities in order.

Reply:
I think that makes sense, a sorted container seems easy enough to add. The rocksdb would also need to be adapted, but it shouldn't be too hard.

Reply:
I did not implement that. But I still think it would be great.

Reply:
Leave a comment to get it improved in a future PR (possibly the sync-v2 PR).

Reply:
Done.

Reply:
Thanks!